dali/python/nvidia/dali/_multiproc/shared_queue.py

```diff
@@ -0,0 +1,233 @@
+        self.shm = shared_mem.SharedMem.allocate(self.shm_capacity)
+        self.is_closed = False
+        self.init_offsets()
+        self.write_meta()
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state['msgs_offsets'] = None
+        state['shm'] = None
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self.init_offsets()
+
+    def init_offsets(self):
+        self.msgs_offsets = [i * self.msg_capacity + self.meta_capacity for i in range(self.capacity)]
+
+    def set_shm(self, shm):
+        assert self.shm is None and shm.capacity >= self.shm_capacity
+        self.shm = shm
+
+    def read_meta(self):
+        self.meta.unpack_from(self.shm.buf, 0)
+
+    def write_meta(self):
+        self.meta.pack_into(self.shm.buf, 0)
+
+    def read_msg(self, i):
+        offset = self.msgs_offsets[i]
+        msg = self.MSG_CLASS()
+        msg.unpack_from(self.shm.buf, offset)
+        return msg
+
+    def write_msg(self, i, msg):
+        offset = self.msgs_offsets[i]
+        msg.pack_into(self.shm.buf, offset)
+
+    def seal(self):
+        self.shm.seal()
+
+    def close(self):
+        if self.is_closed:
+            return
+        with self.lock:
+            self.read_meta()
+            self.is_closed = True
+            if not self.meta.is_closed:
+                self.meta.is_closed = 1
+                self.write_meta()
+                # Notify only one waiting worker about closing of the queue, the woken up worker
+                # will notify the next one. Avoid notify_all at this point, due to possible deadlock if
+                # one of the notified workers exited abruptly when waiting on cv_empty without proper releasing
+                # of the underlying semaphore.
+                self.cv_empty.notify()
+
+    def put(self, msgs : List[MSG_CLASS]) -> Optional[int]:
+        if self.is_closed or not msgs:
+            return
+        with self.lock:
+            self.read_meta()
+            assert self.meta.size + len(msgs) <= self.meta.capacity, "The queue is full"
+            if self.meta.is_closed:
+                self.is_closed = True
+                return
+            msgs_len = len(msgs)
+            next_slot = (self.meta.begining + self.meta.size) % self.meta.capacity
+            for msg in msgs:
+                self.write_msg(next_slot, msg)
+                next_slot = (next_slot + 1) % self.meta.capacity
+            self.meta.size += msgs_len
+            self.write_meta()
+            self.cv_empty.notify()
+        return msgs_len
+
+    def _recv_samples(self, num_samples):
+        num_take = self.meta.size
+        if num_samples is not None and num_samples < num_take:
+            num_take = num_samples
+        recv = [self.read_msg((self.meta.begining + i) % self.meta.capacity) for i in range(num_take)]
+        self.meta.size -= num_take
```

Comment:
```suggestion
        self.meta.size = 0
```
Or it is possible we won't read everything?